text/0000-project-groups.md

```diff
@@ -0,0 +1,328 @@
+much documentation on exactly what a working group does, what are its
+responsibilities, and also importantly setting expectations or goals. There is
+one definition of "Working Group" available in [rust-lang#54445]; quoted below.
+
+> A Rust Working Group is a set of people working at common purpose. Working
+> Groups are associated with a Rust team. Unlike a Rust Team, Working Groups
+> don't have "formal decision making power", though often they are charged with
+> drawing up recommendations, RFCs, or other documents for the teams (which is
+> then intended to make the final decision).
+
+While this definition is true in the broad sense, it does not define the
+difference between what has come to be called "Domain Working Groups" and
+"Team Working Groups". Additionally the community has also adopted the Working
+Group terminology for organising and working on areas they are interested in,
+independent of The Rust Programming Language's Working Groups. It's great that
+we have a community able to self organise in this way, however it has led to
+some confusion over who is supporting these efforts, and whether they're
+considered _"official"_ Working Groups.
+
+This RFC aims to provide clarity by providing new distinct terminology that
+matches the intent behind "Team Working Groups", as well as codify some of the
+processes that these groups have been using to help facilitate creating
+new groups.
+
+[rust-lang#54445]: https://github.com/rust-lang/rust/issues/54445
+
+# Guide-level explanation
+
+To address this confusion this RFC proposes switching from using "Team Working
+Group" in favour of "Project Group". This would serve as a catch all term for
+the sub teams of the existing teams to organise around specific efforts, such
+as certain project or effort in the Rust team. Project Groups are led
+by "Shepherds".
+
+## Life-cycle of a Project Group
+
+This is a high level overview of the complete process of a project group. While
+the flow is built around project groups, we expect that working groups would
+follow a similar process with only minor specifics changed. E.g. A working
+group does not have to find a liaison.
+
+<p align="center">
+    <img src="../resources/project-group-workflow.png"
+        alt="A flow diagram showing each step of a project group"
+        height="800px">
+    <p align="center">Figure 1. Project Group Lifecycle</p>
+</p>
+
+### Steps
+
+1. Exploratory period.
+
+- Initial discussions of the problem area.
+- Write a short motivation for the project.
+- Find a person from the relevant team who's willing to act as a liaison.
+  - Typically you can find someone by creating a post on [internals] or
+    pinging specific people from team to gauge their interest.
+  - You may not always be able to find someone who is willing to act as
+    liasion. It's up to each team to decide how many new efforts
+    they'll have the bandwidth for, which may at times be none.
```

Comment:
I feel uncertain about the procedure laid out here. Intrinsically, there are far fewer people on the team compared to the Rust community at large. I think this procedure is too eager to encourage pinging folks from the teams, who are usually quite busy.

However, that said, this is pretty much an unsolved (and maybe even unsolveable) problem -- how do we dedicate enough time to proposals to sufficiently respond to them in a relatively timely manner, without overloading team members? Maybe one answer here is to ask that people, once they feel like their proposal summary is in a good state, open an issue on some dedicated repository, with the explicit clarification that teams will be asked to respond in "at least" a monthly fashion. This would help set expectations.

On the other hand, maybe I'm creating a problem where there is none, and in practice we expect a much smaller volume of requests here than I'm envisioning.

---

I definitely think we should remove the bit about creating a post on internals though -- or at least we'd want to add a dedicated tag or something -- asking team members to monitor internals today is just not really feasible, it's way too high traffic with too little in the nature of filtering I think.

Reply:
I don't think setting expectations for how quickly the teams would respond would be helpful. It would in my view be too much detail management of the teams, which can have quite varying degrees of bandwidth. The language team will probably get a large number of requests which we don't have time for (and even if we did, I would still want us not to accept proposals to avoid overburdening the compiler team, etc.).

Reply:
Hm, so I think setting some relatively large timeframe is reasonable. For example, a month seems sufficient time for the team to determine for any given request:

1. We do not have time (having not reached this request in any triage process)
1. We did discuss, and are not currently interested -> close issue
1. We did discuss, and are interested, but do not have bandwidth -> leave open
1. We did discuss, and are interested, please expect communication from XXX soon.

Now, having said that, I expect that in many cases option 1 would be selected (especially for relatively higher volume teams, e.g. lang). It could be said that such responses are simply busywork, and in some sense they are, but I think if we accept this RFC we should expect that teams are somewhat actively triaging incoming project group requests, or delegating that activity to some subset.

As I mentioned in another comment, I agree that bandwidth limitations are a fundamental problem. But I think that we should be explicit (and perhaps on a per-team basis!) what the timeframe for responses to requests should look like. If we think a month is too fast, and it may be for some teams, then we can lengthen that, but I think that we should set *some expectation* in this RFC.

Reply:
A month doesn't seem sufficient to me for the language team (looking at https://github.com/rust-lang/rfcs/pulls?q=is%3Aopen+is%3Apr+label%3AT-lang -- there's a lot of proposals coming in), considering that for any given request, we'd need to actually spend some non-trivial time on it. Some of that time will be devoted to resolving disagreements within the team, and this would be time that is taken from e.g., triaging nominated issues, discussing soundness holes, etc. (which I think should be the priorities -- quality assurance for existing stuff, before new features, basically, although not all PGs will be about new features, but mostly, they will, for T-lang).

It seems however fine for some teams to set "you can expect to hear from us in X time", but I think that is best left up to those teams to decide, and not a time-frame set in this RFC for all teams.

Reply:
I agree that the time should be per-team; I'm saying that this RFC should say that each team will record on forge or somewhere that project group formation requests to them will be met with some response within X months.

I agree that there's a high request volume to the language team, and likely other teams as well, but AFAICT, there's no need to resolve disagreements within the team for the initial step we're discussing here. In particular, note that this is specifically about finding a liaison within the team.

To take the lang team, my expectation of how this would work is that each "inactive" -- i.e. not currently liaison-ing -- member of the lang team dedicates 10-15 minutes per month to scan new proposals for discussion groups, and *if* something catches their eye as something they'd like to sponsor, they bring it up at a triage meeting or post on Zulip (this should be quick, just announcing their intent essentially), and then get together with the proposal authors to work on fleshing out their ideas into RFC format to then bring to the lang team.

My assumption is that then there would be *at most* 7 ongoing proposals -- and likely much less -- to the language team. I think that's manageable.

While writing this up, I think I arrived at what may be our disagreement -- you feel that once the language team is "at capacity" you would not want to respond to new project proposals. I agree; we would perhaps make an explicit note that "currently the language team is not available for proposals due to being at capacity" and that would mean that new proposals, if they are opened, are simply closed without review.

Reply:
Sure, it does seem reasonable for each time to record on forge or someplace else will be met in X time with a response. Making "we're at capacity and won't be responding" an option as a "blanket response" would assuage my concerns for the language team. Since this is about finding a liaison, it's also less of a concern to begin with whereas it would have been more pressing if it were about the next point (2. "Obtain the consensus...").

10-15 minutes per month for "inactive" liasons does seem like something each member of the language team, keeping with that running example, is already devoting, so there should be no problem there.

> My assumption is that then there would be _at most_ 7 ongoing proposals -- and likely much less -- to the language team. I think that's manageable.
> 
> While writing this up, I think I arrived at what may be our disagreement -- you feel that once the language team is "at capacity" you would not want to respond to new project proposals. I agree; we would perhaps make an explicit note that "currently the language team is not available for proposals due to being at capacity" and that would mean that new proposals, if they are opened, are simply closed without review.

Just to elaborate a bit to help in understanding each other, I think you do capture the practical outcome of what I think here, but I would add that while I think we might be able to handle 7 ongoing proposals, that would also result in the language team producing too many designs into the compiler team's pipeline, and I think it's best to be conservative with that productivity at this juncture (due to the technical debt, unfinished features, etc.).